### PR TITLE
Error handling and reporting

### PR DESCRIPTION
**Pull request by WardCunningham**
We throw, catch and report errors from algorithmic markdown. They look like this:

image

We catch and report errors thrown by viz.es.js. These look like:

image

### CODE REVIEW

**Comment by WardCunningham** [comment]
I appreciate the refactoring which follows the prescription: make it; make it work, make it right.

File: client/graphviz.js [hunk 1]
```diff
@@ -83,34 +89,41 @@
...
             let tree = nest()
             links.map((link) => {
-              if (ir.match(/^LINKS HERE -> NODE/)) {
-                dot.push(`${quote(context.name)} -> ${quote(link)}`)
-              }
+              if (ir.match(/^LINKS HERE (->|--) NODE/)) {
+                let kind = ir.match(/->|--/)
+                dot.push(`${quote(context.name)} ${kind} ${quote(link)}`)
+              } else
               if (ir.match(/^LINKS NODE -> HERE/)) {
```
Comment (dobbs) [inline client/graphviz.js]: digraph only?

File: client/graphviz.js [hunk 2]
```diff
@@ -83,34 +89,41 @@
...
           if (ir.match(/^HERE/)) {
             let tree = nest()
             let page = await get(context)
             if (page) {
               if (ir.match(/^HERE NODE/)) {
                 dot.push(quote(context.name))
-              }
+              } else
               if (ir.match(/^HERE NODE \w+/)) {
                 dot.push(`${quote(ir)} -> ${quote(context.name)} [style=dotted]`)
```
Comment (dobbs) [inline client/graphviz.js]: digraph only?

File: client/graphviz.js [hunk 3]
```diff
@@ -130,23 +143,21 @@
...
-          }
+          } else
 
           if (ir.match(/^FAKE/)) {
-            if (ir.match(/^FAKE HERE -> NODE/)) {
+            if (ir.match(/^FAKE HERE (->|--) NODE/)) {
               dot.push(`${quote(context.name)} -> ${quote('post-'+context.name)}`)
-            }
+            } else
             if (ir.match(/^FAKE NODE -> HERE/)) {
```
Comment (dobbs) [inline client/graphviz.js]: digraph only?

File: client/graphviz.js (hunk 1 above)
Comment (WardCunningham) [reply client/graphviz.js]: Here I correctly capture `kind` and substitute it into the dot format instruction. This is the case I tested.

File: client/graphviz.js (hunk 2 above)
Comment (WardCunningham) [reply client/graphviz.js]: You are right. Here and everywhere else we emit dot format links we need to remember whether or not we were asked to generate a directed graph. I will make sure this information is recorded in the context and appropriately substituted in the places you have identified. Commit coming soon.

**Comment by WardCunningham** [comment]
I have resolved the correct usage of undirected links when required as @dobbs has noticed. To my surprise the direction of links in an undirected graph does make a difference. Here are two versions of the same graph changing `HERE -- NODE` to `NODE -- HERE`:

image

image

Graphviz appears to be making this distinction based solely on the order in which it encounters names in its input. It is good that we can produce either.

File: client/graphviz.js (hunk 2 above)
Comment (WardCunningham) [reply client/graphviz.js]: Commit has arrived. See 4ac631e.

**Comment by WardCunningham** [comment]
Our commit 4ac631e incorrectly detected the need for or absence of directional edges for source labeling (dotted lines) of HERE nodes. Further the HERE NODE version of this case was hidden by an earlier overly aggressive regex. Here we correct both so that we can draw this diagram.
image

**Comment by WardCunningham** [comment]
I've moved the header processing outside of the eval loop mostly anticipating some sort of transclusion encoded into a different form of this header. It also never made sense to evaluate DOT at any intermediate point in the markup.

**Pull request merged**